`text` index: Change default index granularity to 64

Ergus opened the pull request:
### Changelog category (leave one):
- Experimental Feature

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Changed the default index granularity value for `text` indexes to 64. This improves the expected performance for the average test query in internal benchmarks.

Ergus committed:
Set text index default GRANULARITY to 64

Given the results in our benchmarks 64 provided better performance.
So, we use this value until further optimizations are implemented.

Ergus committed:
Update documentation

rschu1ze commented on docs/en/engines/table-engines/mergetree-family/invertedindexes.md:
```diff
@@ -35,26 +35,26 @@ useful the resulting full-text index will be.
     `key` UInt64,
     `str` String,
-    INDEX inv_idx(str) TYPE text(tokenizer = 'default|ngram|split|no_op' [, ngram_size = N] [, separators = []] [, max_rows_per_postings_list = M]) GRANULARITY 1
+    INDEX inv_idx(str) TYPE text(tokenizer = 'default|ngram|split|no_op' [, ngram_size = N] [, separators = []] [, max_rows_per_postings_list = M]) GRANULARITY 64
```
Could you wrap `GRANULARITY 64` in `[` / `]`. This would give readers the impressino that specifying the granularity is optional.

Ergus committed:
Force backward compatibility by simplifying

Fixes tests: 02346_text_index_legacy_name and 02346_text_index_experimental_flag

Ergus commented on docs/en/engines/table-engines/mergetree-family/invertedindexes.md:
```diff
@@ -1,22 +1,22 @@
+sidebar_label: 'text Indexes'
 slug: /engines/table-engines/mergetree-family/invertedindexes
-title: 'Full-text Search using Full-text Indexes'
+title: 'Full-text Search using text Indexes'
```
Done:

However I was actually wondering if some sort of quotation would be better, as tests in this case is a reserved keyword

Ergus committed:
Fix typos and small improve in documentation

Ergus committed:
Add test for the changes

* tests/queries/0_stateless/02346_text_index_default_granularity.{sql,reference} : New files with test

Ergus commented:
> 02354_vector_search_default_granularity.sql

Done

rschu1ze commented on docs/en/engines/table-engines/mergetree-family/invertedindexes.md:
```diff
@@ -35,26 +35,26 @@ useful the resulting full-text index will be.
     `key` UInt64,
     `str` String,
-    INDEX inv_idx(str) TYPE text(tokenizer = 'default|ngram|split|no_op' [, ngram_size = N] [, separators = []] [, max_rows_per_postings_list = M]) GRANULARITY 1
+    INDEX inv_idx(str) TYPE text(tokenizer = 'default|ngram|split|no_op' [, ngram_size = N] [, separators = []] [, max_rows_per_postings_list = M]) [GRANULARITY 64]
```
Minor: I think instead of `64` we want a placeholder here, e.g. `GRANULARITY gran`. L. 80 can then reference the placeholder.

rschu1ze commented:
@ahmadov FYI

Ergus commented:
The error in the failing tests `02346_text_index_queries` seems totally unrelated with this change. The problem is accessing the `system.query_log` as you can see in the error logs.

It looks like an issue in the master branch over which I implemented this. I tried to replicate locally but the `system.query_log` table is not created.  I tried the config file:

```
<?xml version="1.0"?>
<clickhouse>

    <query_log>
        <database>system</database>
        <table>query_log</table>
        <engine>Engine = MergeTree PARTITION BY event_date ORDER BY event_time TTL event_date + INTERVAL 30 day</engine>
        <flush_interval_milliseconds>7500</flush_interval_milliseconds>
    </query_log>

</clickhouse>
```

and starting the server with `-C. config.xml` but that trigger an error:

```
2025.06.20 00:01:28.663668 [ 31804 ] {} <Error> Application: Code: 76. DB::ErrnoException: Cannot open file /var/lib/clickhouse/status: , errno: 2, strerror: No such file or directory. (CANNOT_OPEN_FILE), Stack trace (when copying this message, always include the lines below):

0. /home/jimmy/storage/repos/ClickHouse/contrib/llvm-project/libcxx/include/__exception/exception.h:113: std::exception::capture() @ 0x000000000f68a9c2
1. /home/jimmy/storage/repos/ClickHouse/contrib/llvm-project/libcxx/include/__exception/exception.h:85: std::exception::exception[abi:se190107]() @ 0x000000000f68a98d
2. /home/jimmy/storage/repos/ClickHouse/base/poco/Foundation/src/Exception.cpp:27: Poco::Exception::Exception(String const&, int) @ 0x000000002cbd3640
3. /home/jimmy/storage/repos/ClickHouse/src/Common/Exception.cpp:110: DB::Exception::Exception(DB::Exception::MessageMasked&&, int, bool) @ 0x00000000199f983e
4. /home/jimmy/storage/repos/ClickHouse/src/Common/Exception.h:119: DB::Exception::Exception(String&&, int, bool) @ 0x000000000f685bf7
5. /home/jimmy/storage/repos/ClickHouse/src/Common/Exception.h:57: DB::Exception::Exception(PreformattedMessage&&, int) @ 0x000000000f6833a9
6. /home/jimmy/storage/repos/ClickHouse/src/Common/Exception.h:242: DB::ErrnoException::ErrnoException<String const&>(int, int, FormatStringHelperImpl<std::type_identity<String const&>::type>, String const&) @ 0x0000000019b5ac81
7. /home/jimmy/storage/repos/ClickHouse/src/Common/Exception.h:257: void DB::ErrnoException::throwFromPath<String const&>(int, String const&, FormatStringHelperImpl<std::type_identity<String const&>::type>, String const&) @ 0x0000000019b5ab3a
8. /home/jimmy/storage/repos/ClickHouse/src/Common/StatusFile.cpp:67: DB::StatusFile::StatusFile(String, std::function<void (DB::WriteBuffer&)>) @ 0x0000000019ef25ea
9. /home/jimmy/storage/repos/ClickHouse/programs/server/Server.cpp:1380: DB::Server::main(std::vector<String, std::allocator<String>> const&) @ 0x0000000019f6299f
10. /home/jimmy/storage/repos/ClickHouse/base/poco/Util/src/Application.cpp:315: Poco::Util::Application::run() @ 0x000000002ccff292
11. /home/jimmy/storage/repos/ClickHouse/programs/server/Server.cpp:594: DB::Server::run() @ 0x0000000019f59434
12. /home/jimmy/storage/repos/ClickHouse/base/poco/Util/src/ServerApplication.cpp:131: Poco::Util::ServerApplication::run(int, char**) @ 0x000000002cd1f905
13. /home/jimmy/storage/repos/ClickHouse/programs/server/Server.cpp:391: mainEntryClickHouseServer(int, char**) @ 0x0000000019f56b4b
14. /home/jimmy/storage/repos/ClickHouse/programs/main.cpp:321: main @ 0x000000000f66a6c2
15. ? @ 0x00000000000276b5
16. ? @ 0x0000000000027769
17. _start @ 0x000000000f64f02e
 (version 25.6.1.1)
 ```
 
 I could be doing something wrong, if so, just point me in the right direction.
 
 else 
 
 As I don't see an open issue related to this error, but a couple of fixes today related with logging. So, I will try to reproduce on master and if I can create a `system.query_log` there I will rebase this branch over it and test again 
 
 else 
 
 May I open a new issue and work on this?

ahmadov reviewed (approved):
LGTM 👍

Ergus committed:
Fix failing test: 02346_text_index_queries

This sincerely looks like a workaround until the dichotomy between `GRANULARITY` and `SETTINGS index_granularity = ` is solved

rschu1ze commented:
@Ergus GRANULARITY and index_granularity are different things and they are named confusingly, sorry.

The granularity (or table granularity) is a table-level property which is controlled by setting `index_granularity`. It controls the number of rows per granule (8192 by default).

The index granularity is an index-level property which is controlled by keyword GRANULARITY. It controls the number of (table) granules per index granules. More details are [here](https://clickhouse.com/docs/optimize/skipping-indexes).

I know this is confusing but in an ideal world, users don't need to tune either parameter.

Integration tests (release, 3/4)
- test_keeper_invalid_digest: Timeout. Seems [extremely sporadic](https://play.clickhouse.com/play?user=play#U0VMRUNUIHB1bGxfcmVxdWVzdF9udW1iZXIgYXMgcHIsIGNoZWNrX3N0YXJ0X3RpbWUsIGNoZWNrX25hbWUsIHRlc3RfbmFtZSwgdGVzdF9zdGF0dXMsIHJlcG9ydF91cmwtLSwgdGVzdF9jb250ZXh0X3JhdwpGUk9NIGNoZWNrcwpXSEVSRSAxCiAgICAtLUFORCBwdWxsX3JlcXVlc3RfbnVtYmVyID0gMAogICAgQU5EIHRlc3Rfc3RhdHVzIElOICgnRkFJTCcsICdFUlJPUicsICdCUk9LRU4nKQogICAgQU5EIHRvRGF0ZShjaGVja19zdGFydF90aW1lKSA+ICcyMDI1LTAxLTAxJwogICAgQU5EIHRlc3RfbmFtZSBsaWtlICcldGVzdF9rZWVwZXJfaW52YWxpZF9kaWdlc3QlJwogICAgLS0gQU5EIHRlc3RfY29udGV4dF9yYXcgbGlrZSAnJUFMVEVSX01FVEFEQVRBJScKT1JERVIgQlkgY2hlY2tfc3RhcnRfdGltZSBERVNDCgo=) (probably caused by infra) and I'll not open a new issue.

Perf tests:
- Also ran into weird infra errors. Probably fixed in the meantime. The perf tests don't check text indexes, so we are good.
